### PR TITLE
WT-7524 Refactor functions to obtain checkpoint list; Clear delete on skipping checkpoints

### PR DESCRIPTION
**Pull request by sulabhM**

### CODE REVIEW

**Comment by sulabhM** [comment]
@keithbostic , please have a look. This change brings back a single function to get the checkpoint list. It also resolves a couple of bugs I identified, while making this change.

File: src/txn/txn_ckpt.c [hunk 1]
```diff
@@ -1318,8 +1318,13 @@ __checkpoint_lock_dirty_tree_int(WT_SESSION_IMPL *session, bool is_checkpoint, b
...
      */
     WT_RET(__checkpoint_mark_skip(session, ckptbase, force));
-    if (F_ISSET(btree, WT_BTREE_SKIP_CKPT))
+    if (F_ISSET(btree, WT_BTREE_SKIP_CKPT)) {
+        WT_CKPT_FOREACH (ckptbase, ckpt)
+            if (F_ISSET(ckpt, WT_CKPT_DELETE))
+                F_CLR(ckpt, WT_CKPT_DELETE);
         return (0);
+    }
+
```
Comment (keithbostic) [inline src/txn/txn_ckpt.c:1327]: @sulabhM, this seems like an orthogonal change, can you please explain what's going on here?

File: src/txn/txn_ckpt.c [hunk 2]
```diff
@@ -1539,7 +1534,8 @@ __checkpoint_lock_dirty_tree(
         /* It is possible that we do not have any checkpoint in the list. */
 err:
         __wt_meta_ckptlist_free(session, &ckptbase);
-        __wt_meta_saved_ckptlist_free(session);
+        btree->ckpt = NULL;
+        btree->ckpt_bytes_allocated = 0;
```
Comment (keithbostic) [inline src/txn/txn_ckpt.c:1538]: @sulabhM, this seems like an orthogonal change, can you please explain what's going on here?

Comment (keithbostic) [reply src/txn/txn_ckpt.c:1538]: I understand this change now, never mind.

File: src/txn/txn_ckpt.c (hunk 1 above)
Comment (sulabhM) [reply src/txn/txn_ckpt.c:1327]: It is a bug that started showing in the tests with these changes that I made. In some cases, we decide to skip checkpointing a btree. In such cases, we should clear the "delete" flag on the checkpoints that were to be deleted. If we do not do so, the checkpoints stay marked as deleted in the saved list. The next time we checkpoint and use the saved list, the debug code tries to generate the checkpoint list from the metadata to check for a match. There a mismatch in the flags show up.

I think this is a bug in the former code, for some reason the testing doesnt uncover it.

Comment (sulabhM) [reply src/txn/txn_ckpt.c:1332]: I have added a comment explaining the reason behind clearing delete flags on checkpoints

**Review by etienneptl — APPROVED** [review]
LGTM!

File: src/txn/txn_ckpt.c (hunk 1 above)
Comment (keithbostic) [reply src/txn/txn_ckpt.c:1332]: > I think this is a bug in the former code, for some reason the testing doesn't uncover it.

Makes sense -- would it be safer to clear the delete flag when the checkpoint list is acquired, in all cases, rather than only clear the flag when the skip happens? Does that make it less likely we'll introduce future bugs? (To be clear, leaving this entirely up to you, I'm just asking the question in case you like the idea.)

**Review by keithbostic — APPROVED** [review]
lgtm

File: src/txn/txn_ckpt.c (hunk 1 above)
Comment (sulabhM) [reply src/txn/txn_ckpt.c:1332]: That is not a bad idea. I feel hesitation leaving the checkpoints marked as to be deleted in the cached list, after having skipped a checkpoint. The debug code we have put tries to match cached checkpoint's flags with the ones generated from metadata. We should be able to catch such issues there.

I am going to get a MongoDB patch test going and if green merge for now. I will keep your suggestion handy if we need to go there in the future.
